File: lib/models/plex_sort.dart
import 'package:json_annotation/json_annotation.dart';

part 'plex_sort.g.dart';

@JsonSerializable()
class PlexSort {
  final String key;
  final String? descKey;
  final String title;
  final String? defaultDirection;

  PlexSort({required this.key, this.descKey, required this.title, this.defaultDirection});

  factory PlexSort.fromJson(Map<String, dynamic> json) => _$PlexSortFromJson(json);

  Map<String, dynamic> toJson() => _$PlexSortToJson(this);

  /// Gets the full sort key with direction
  /// If [descending] is true, returns the descKey or key:desc
  /// Otherwise returns the key for ascending sort
  String getSortKey({bool descending = false}) {
    if (!descending) {
      return key;
    }

    // Use descKey if available, otherwise append :desc to key
    return descKey ?? '$key:desc';
  }

  /// Returns true if this sort's default direction is descending
  bool get isDefaultDescending {
    return defaultDirection?.toLowerCase() == 'desc';
  }

  @override
  String toString() {
    return 'PlexSort(key: $key, title: $title, defaultDirection: $defaultDirection)';
  }

  @override
  bool operator ==(Object other) {
    if (identical(this, other)) return true;
    return other is PlexSort && other.key == key;
  }

  @override
  int get hashCode => key.hashCode;
}

File: apple-tv/VibeStream/Models/PlexSort.swift
import Foundation

struct PlexSort: Identifiable, Hashable {
    let key: String
    let title: String
    var descKey: String?
    var defaultDirection: String?

    var id: String { key }

    var isDefaultDescending: Bool { defaultDirection == "desc" }

    func sortKey(descending: Bool) -> String {
        if descending {
            return descKey ?? "\(key):desc"
        }
        return key
    }

    static func from(json: [String: Any]) -> PlexSort? {
        guard let key = json["key"] as? String,
              let title = json["title"] as? String else {
            return nil
        }
        return PlexSort(
            key: key,
            title: title,
            descKey: json["descKey"] as? String,
            defaultDirection: json["defaultDirection"] as? String
        )
    }
}
